fix Yasharn, Implacable Earth and Angel of Jubilation

Jmlundeen opened the pull request:
This fixes Angel of Jubilation and Yasharn, Implacable Earth incorrectly allowing/preventing sacrifice costs. Angel of Jubilation was blocking sacrifice costs for triggered abilities and Yasharn was not preventing sacrifice and life costs.

I fixed some of the Yasharn tests and added a couple of new tests to each. These tests should cover most scenarios. I created a common effect and just replaced the logic in Yasharn, so I figured it should be a PR first.

closes #9650
closes #13422

Jmlundeen commented:
[[Yasharn, Implacable Earth]] [[Angel of Jubilation]]

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/CantPayLifeOrSacrificeEffect.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    public CantPayLifeOrSacrificeEffect(FilterPermanent filter) {
+        super(Duration.WhileOnBattlefield, Layer.RulesEffects, SubLayer.NA, Outcome.Detriment);
+        this.filter = filter;
```
use `filter.copy()` since you modify it

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/CantPayLifeOrSacrificeEffect.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public boolean apply(Game game, Ability source) {
+        for (UUID playerId : game.getState().getPlayersInRange(source.getControllerId(), game)) {
+            Player player = game.getPlayer(playerId);
```
miss null check

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/CantPayLifeOrSacrificeEffect.java:
```diff
@@ -0,0 +1,52 @@
+    public boolean apply(Game game, Ability source) {
+        for (UUID playerId : game.getState().getPlayersInRange(source.getControllerId(), game)) {
+            Player player = game.getPlayer(playerId);
+            player.setPayLifeCostLevel(Player.PayLifeCostLevel.nonSpellnonActivatedAbilities);
```
The underlying logic is a bad approach that can cause bugs.

Test case: cast Yasharn, then cast Karn's Sylex. Try to pay life to activate a mana ability. It should be prevented by Yasharn, but seems like Karn's Sylex will apply afterwards (timestamp), so with current implementation, it would not be prevented.

Since only two options, I would remove `PayLifeCostLevel` entirely and just have two booleans one after another `canPayLifeToCastSpellsActivateAbilities` and `canPayLifeToCastSpellsActivateNonManaAbilities`. (A more extensible approach for future could be an Enum `PayLifeCostRestriction` `SPELLS`, `NON_MANA_ABILITIES`, `MANA_ABILITIES`, with an EnumSet stored in PlayerImpl (cleared on reset), and the effect adds the appropriate items to the set for each player in range.)

xenohedron commented on Mage/src/main/java/mage/abilities/effects/common/CantPayLifeOrSacrificeEffect.java:
```diff
@@ -0,0 +1,52 @@
+        for (UUID playerId : game.getState().getPlayersInRange(source.getControllerId(), game)) {
+            Player player = game.getPlayer(playerId);
+            player.setPayLifeCostLevel(Player.PayLifeCostLevel.nonSpellnonActivatedAbilities);
+            player.setCanPaySacrificeCostFilter(filter);
```
The underlying logic is another bad approach that can cause bugs.

Test case: cast Yasharn, then cast Angel of Jubilation. Try to sacrifice a noncreature nonland permanent to activate ability. It should be prevented by Yasharn, but seems like Angel of Jubilation will apply afterwards (timestamp), so with current implementation, it would not be prevented.

I don't love having this filter logic in PlayerImpl. I wonder if the existing `canPaySacrificeCost` could instead try a new `GameEvent.PAY_SACRIFICE_COST` that's checked by a ContinuousRuleModifyingEffect. (And then you wouldn't have to add a custom predicate to the filter in order to handle that part.)

Jmlundeen replied:
This also makes sense, not really setup for layering. Shouldn't be too difficult to change that? Maybe the life restriction could do something like this as well? I'd have to look at the underlying code, but it would be a bit neater to have these just be one rule modifying effect instead of the continuous and a rule modify.

xenohedron commented on Mage/src/main/java/mage/filter/predicate/mageobject/SpellOrActivatedAbilitySourcePredicate.java:
```diff
@@ -0,0 +1,21 @@
+    public String toString() {
+        return "Source is a spell or activated ability";
+    }
+}
```
nit, can you get your IDE to automatically include exactly one trailing newline at end of file?

Jmlundeen replied:
Is there a shared code style scheme for java for IntelliJ? That would be good to have to keep everyone consistent. Mine's just running stock with changes to imports i think.

JayDi85 replied:
Just use default intellij settings. No need in additional setup. Empty line at the end is not required)) Just make sure git auto-replace r/rn line ending (wiki page has instructions about it).

P.S. How to setup empty line: https://stackoverflow.com/a/16761228

JayDi85 replied:
Maybe max line size can be changed from 70 to 140+. I’m not sure is it require for new IntelliJ IDEA setup, but vsc enable it by default and that’s big pain on every language and project.

JayDi85 replied:
Yes, it is

xenohedron commented on Mage/src/main/java/mage/filter/predicate/mageobject/SpellOrActivatedAbilitySourcePredicate.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public String toString() {
```
is this actually used for anything on a predicate?

xenohedron commented on Mage.Sets/src/mage/cards/a/AngelOfJubilation.java:
```diff
@@ -26,6 +21,8 @@
  */
 public final class AngelOfJubilation extends CardImpl {
 
+    private static final FilterCreaturePermanent filter = new FilterCreaturePermanent("creatures");
```
StaticFilters.FILTER_PERMANENT_CREATURES

(if you were trying to avoid using it because of the effect modifying it, well, you should use copy() in that case anyway)

(same for Yasharn)

xenohedron commented on Mage.Sets/src/mage/cards/a/AngelOfJubilation.java:
```diff
@@ -52,67 +48,4 @@ private AngelOfJubilation(final AngelOfJubilation card) {
-    }
-}
-
-class AngelOfJubilationSacrificeFilterEffect extends CostModificationEffectImpl {
```
so this part of the implementation was useless because the PayLifeCostLevel was already handling it?

Jmlundeen replied:
This part seemed to be aimed at the sacrifice cost, but that was already *mostly* being handled by the canPay. This doesn't seem like the correct way to handle the sacrifice cost though and caused issues when I tried it with Yasharn.

Jmlundeen commented:
Ok, I created the new `PAY_SACRIFICE_COST` event and updated `canPaySacrificeCost` to use it. I also converted the effect into an ability to combine both the cant pay life and cant pay sacrifice costs into one. This works correctly now if you have Karn's Sylex + Yasharn/Jubilation and I added test coverage.

I also updated Thran Portal since it wasn't applying the pay life cost correctly.

xenohedron commented on Mage/src/main/java/mage/abilities/common/CantPayLifeOrSacrificeAbility.java:
```diff
@@ -0,0 +1,150 @@
+ */
+public class CantPayLifeOrSacrificeAbility extends SimpleStaticAbility {
+
+    final String rule;
```
private

xenohedron commented on Mage/src/main/java/mage/abilities/common/CantPayLifeOrSacrificeAbility.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * @param nonManaAbilities boolean to set if the restriction should only apply to non-mana abilities
+     */
+    public CantPayLifeEffect(boolean nonManaAbilities) {
```
not public (package-private)

xenohedron commented on Mage/src/main/java/mage/players/Player.java:
```diff
@@ -57,16 +53,13 @@
-    enum PayLifeCostLevel {
-        allAbilities, nonSpellnonActivatedAbilities, onlyManaAbilities, none
+    enum PayLifeCostRestriction {
+        CAST_SPELLS, ACTIVATE_NON_MANA_ABILITIES, ACTIVATE_ABILITIES
```
So I like the enum set approach. However, I do think it would be cleaner if they were all mutually exclusive. Instead of "activate abilities" have "activate mana abilities".

xenohedron commented on Mage/src/main/java/mage/abilities/common/CantPayLifeOrSacrificeAbility.java:
```diff
@@ -0,0 +1,150 @@
+    private final FilterPermanent sacrificeFilter;
+    private final boolean allowNonManaAbilities;
+
+    public CantSacrificeEffect(boolean allowNonManaAbilities, FilterPermanent sacrificeFilter) {
```
not public

xenohedron commented on Mage/src/main/java/mage/abilities/common/CantPayLifeOrSacrificeAbility.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    @Override
+    public ContinuousEffect copy() {
```
CantSacrificeEffect (covariant return)

xenohedron commented on Mage.Sets/src/mage/cards/b/BiteDownOnCrime.java:
```diff
@@ -65,7 +65,7 @@ enum BiteDownOnCrimeAdjuster implements CostAdjuster {
     public void reduceCost(Ability ability, Game game) {
         if (CollectedEvidenceCondition.instance.apply(game, ability)
-                || (game.inCheckPlayableState() && collectEvidenceCost.canPay(ability, null, ability.getControllerId(), game))) {
+                || (game.inCheckPlayableState() && collectEvidenceCost.canPay(ability, ability, ability.getControllerId(), game))) {
```
can you just remind me what problem this change solves?

Jmlundeen replied:
This is so the canPaySacrificeCost is able to check the source ability to determine if a restriction should be applied. These bargain cards seemed to be the only ones passing a null source.

JayDi85 replied:
For info: canPay can call canTarget or possibleTargets inside — so null source ability will “disable” color and other target protection checks - and that’s correct cause choosing to pay is not same as choosing target and must ignore it. Nullable source was used as alternative to target.setNotTarget modification (setNotTarget must be priority today).

xenohedron commented on Mage/src/main/java/mage/abilities/common/CantPayLifeOrSacrificeAbility.java:
```diff
@@ -0,0 +1,150 @@
+
+    /**
+     * @param allowNonManaAbilities boolean to set if the restriction should only apply to non-mana abilities
+     * @param sacrificeFilter filter for types of permanents that cannot be sacrificed
```
add comment about null usage

(another approach could be to call it CantPayLifeAbility, with a chaining method orSacrifice)

xenohedron reviewed (approved):
sorry for the delayed review, a few nits but overall looks good to me

JayDi85 commented:
@Jmlundeen can you look at #13901? Maybe it can be related to current PR
